client/src/components/room/streamGate.tsx

```diff
@@ -0,0 +1,136 @@
+import {
+  Call,
+  StreamCall,
+  StreamVideo,
+  StreamVideoClient,
+  User,
+} from "@stream-io/video-react-sdk";
+
+import { useCallback, useEffect, useState } from "react";
+
+import { logError, logInfo } from "../../lib/logger";
+import { getStreamAdminCredentials } from "../../lib/server-api";
+import { useAdminStore } from "../../stores/adminStore";
+import { useRoomStore } from "../../stores/roomStore";
+
+const publicUser: User = { type: "anonymous" };
+
+/**
+ * StreamGate provides context for being in a Stream call as an admin or viewer. Upon initializing, it will:
+ *
+ * - (As an admin) Get or create a stream call and admin RTMPS credentials
+ * - (All roles) Join the call as an admin or anonymous viewer
+ *
+ * Upon leaving the page, the user will automatically leave the call.
+ *  */
+const StreamGate: React.FunctionComponent<{
+  children: any;
+  roomID: string;
+  streamPlaybackID?: string;
+  anonymousOnly: boolean;
+}> = ({ children, roomID, streamPlaybackID, anonymousOnly }) => {
+  const isAdmin = useAdminStore(useCallback((s) => s.isAdmin, []));
+
+  const [state, setState] = useState<{
+    client: StreamVideoClient;
+    call: Call;
+    rtmpsDetails: RtmpsDetails | null;
+  }>();
+
+  const useAdmin = !anonymousOnly && isAdmin;
+
+  useEffect(() => {
+    getClient(roomID, useAdmin).then(([myClient, rtmpsDetails]) => {
+      if (!streamPlaybackID) {
+        // Technically, streamPlaybackID may be indirectly set in the middle of getClient()
+        // so there is a race condition here. If this happens, we will call getClient twice,
+        // and that's okay.
+        return;
+      }
+
+      const myCall = myClient.call("livestream", streamPlaybackID);
+
+      myCall
+        .get()
+        .then(() => {
+          logInfo("Setting state for call ID " + streamPlaybackID);
+          setState({ client: myClient, call: myCall, rtmpsDetails });
+        })
+        .catch((e) => logError("Failed to retrieve call details", e));
+    });
+  }, [roomID, streamPlaybackID, useAdmin]);
+
+  useEffect(() => {
+    if (!state) {
+      return;
+    }
+
+    logInfo("....JOINING CALL....");
+
+    // Ensure camera is disabled by default on page load.
+    state.call.camera.disable();
+
+    state.call
+      .join()
+      .then(() => {
+        logInfo("Joined call from StreamGate");
+      })
+      .catch((e) => {
+        logError("Failed to join call", e);
+      });
+
+    return () => {
+      state.call.leave().catch((e) => {
+        logError("Failed to leave call", e);
+      });
+    };
+  }, [state]);
+
+  if (!state || !state.client || !state.call) return null;
+
+  return (
+    <StreamVideo client={state.client}>
+      <StreamCall call={state.call}>{children(state.rtmpsDetails)}</StreamCall>
```

Comment:
It seems like this rtmps details is being used by AdminStreamPanel like 3 levels down. I have previously held a (slight) preference for things to not be based down over multiple layers, and instead use zustand for that sort info. 

Curious why you went this direction! Thought I may learn something by asking :) Also seems odd that MUX doesn't provide that in useCall

Reply:
I was mostly in a hurry to get it to work :) I agree the clean way would have probably been to make a new zustand store for this.

Stream _does_ provide a hook for the RTMP address but it wasn't always reliably populated (sometimes it was null?) The Stream Key is also something that has to be generated by us since it's a short-lived admin key.